DMCompiler/Optimizer/PeepholeOptimizer.cs

```diff
@@ -25,44 +29,58 @@ public static void ReplaceInstructions(List<IAnnotatedBytecode> input, int index
     }
 }
 
+/// <summary>
+/// The list of peephole optimizer passes in the order that they should run
+/// </summary>
+internal enum OptPass : byte {
+    PeepholeOptimization = 1,   // First-pass peephole optimizations (e.g. const folding)
+    BytecodeCompactor = 2,      // Next-pass bytecode compacting (e.g. PushNFloats and other PushN opcodes)
+    ListCompactor = 3           // Final-pass list compacting (e.g. PushNFloats & CreateList -> CreateListNFloats)
+}
+
+// ReSharper disable once ClassNeverInstantiated.Global
 internal sealed class PeepholeOptimizer {
     private class OptimizationTreeEntry {
-        public IPeepholeOptimization? Optimization;
+        public IOptimization? Optimization;
         public Dictionary<DreamProcOpcode, OptimizationTreeEntry>? Children;
     }
 
     /// <summary>
     /// Trees matching chains of opcodes to peephole optimizations
     /// </summary>
-    private static readonly Dictionary<DreamProcOpcode, OptimizationTreeEntry> OptimizationTrees = new();
+    /// <remarks>This is instantiated when the optimizer runs</remarks>
```

Comment:
I think initializing in a static constructor would be better, and wouldn't need to suppress a nullable warning with `null!`

It could also be made an array with a dictionary for each value of `OptPass` which would avoid unnecessarily looping over & creating the same types multiple times.